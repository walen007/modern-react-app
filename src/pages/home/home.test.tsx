import React from 'react';
import { render, waitFor, fireEvent } from '@test/test-utils';
import { pastLaunchesAll } from '@test/past-launches-all';
import { searchLaunchId } from '@test/search';
import { pastLaunchesMixed } from '@test/past-launches-mixed';

const Home = React.lazy(() => import('.'));

const allPastLaunches: Response = {
  ok: true,
  status: 200,
  statusText: 'OK',
  headers: {
    // @ts-ignore
    'Content-Type': 'application/json',
  },
  json: () => Promise.resolve(pastLaunchesAll),
};

vi.mock('./LazyComponent', () => {
  return {
    __esModule: true,
    default: () => <Home />,
  };
});

describe('<Home />', () => {
  afterEach(() => {
    vi.useRealTimers();
  });

  it('--- should render home page lazy', async () => {
    vi.spyOn(window, 'fetch').mockResolvedValueOnce(allPastLaunches);

    const { baseElement, getByText } = render(
      <React.Suspense fallback="Loading home page for testing">
        <Home />
      </React.Suspense>
    );

    await waitFor(() => getByText(/Past launches/));
    await waitFor(() => getByText('ID: 62dd70d5202306255024d139'));

    expect(baseElement).toMatchSnapshot();
  });

  it('--- should render page load error', async () => {
    vi.spyOn(window, 'fetch').mockResolvedValueOnce(searchLaunchId('error'));

    const { getByTestId } = render(
      <React.Suspense fallback="Loading home page for testing">
        <Home />
      </React.Suspense>
    );

    const errorMessage = await waitFor(() => getByTestId('error-message'));
    expect(errorMessage).toHaveTextContent('An error occurred while retrieving launch data.');
  });

  it('--- should render search result (dynamic content) with successful launch indication', async () => {
    vi.spyOn(window, 'fetch').mockResolvedValueOnce(allPastLaunches);
    const { getByText, getByPlaceholderText, getByTestId } = render(
      <React.Suspense fallback="Loading home page for testing">
        <Home />
      </React.Suspense>
    );

    await waitFor(() => getByText('Past launches'));

    const input = getByPlaceholderText('Enter Launch ID');
    const button = getByText('Search');

    fireEvent.change(input, {
      target: { value: pastLaunchesMixed[0].id },
    });

    // Mock system time
    vi.setSystemTime(new Date('2024-02-10T12:00:00Z'));

    vi.spyOn(window, 'fetch').mockResolvedValue(searchLaunchId(pastLaunchesMixed[0].id));
    fireEvent.click(button);

    expect(await waitFor(() => getByText('Search result'))).toBeInTheDocument();
    expect(await waitFor(() => getByText('Elapsed time since launch'))).toBeInTheDocument();

    expect(getByTestId('time-elapsed')).toHaveTextContent('529:00:00');

    const indicator = getByTestId('success-indicator');
    expect(indicator.getAttribute('aria-description')).toBe(
      'Success Indicator: The SpaceX Crew-5 was a successful launch'
    );
  });

  it('--- should render search result (dynamic content) with failed launch indication', async () => {
    vi.spyOn(window, 'fetch').mockResolvedValueOnce(allPastLaunches);
    const { getByText, getByPlaceholderText, getByTestId } = render(
      <React.Suspense fallback="Loading home page for testing">
        <Home />
      </React.Suspense>
    );

    await waitFor(() => getByText('Past launches'));

    const input = getByPlaceholderText('Enter Launch ID');
    const button = getByText('Search');

    fireEvent.change(input, {
      target: { value: pastLaunchesMixed[1].id },
    });

    // Mock system time - to prevent fluctuation in date calculations
    vi.setSystemTime(new Date('2024-02-10T12:00:00Z'));

    vi.spyOn(window, 'fetch').mockResolvedValue(searchLaunchId(pastLaunchesMixed[1].id));
    fireEvent.click(button);

    expect(await waitFor(() => getByText('Search result'))).toBeInTheDocument();
    expect(await waitFor(() => getByTestId('time-elapsed'))).toBeInTheDocument();

    const timeElapsed = getByTestId('time-elapsed');
    expect(timeElapsed).toHaveTextContent('6775:30:0');

    const indicator = getByTestId('success-indicator');
    expect(indicator.getAttribute('aria-description')).toBe(
      'Success Indicator: The SpaceX FalconSat was not a successful launch'
    );
  });

  it('--- should render error state ui', async () => {
    const randomStr = 'asdsasda8aidjnad80audja098d';
    vi.spyOn(window, 'fetch').mockResolvedValueOnce(allPastLaunches);
    const { getByText, getByPlaceholderText, getByTestId } = render(
      <React.Suspense fallback="Loading home page for testing">
        <Home />
      </React.Suspense>
    );

    await waitFor(() => getByText('Past launches'));

    const input = getByPlaceholderText('Enter Launch ID');
    const button = getByText('Search');

    fireEvent.click(button);
    const errorMessage = await waitFor(() => getByTestId('error-message'));
    expect(errorMessage).toHaveTextContent('Please enter a valid Launch ID');

    fireEvent.change(input, {
      target: { value: randomStr },
    });

    // Mock system time - to prevent fluctuations in date calculations
    vi.setSystemTime(new Date('2024-02-10T12:00:00Z'));

    vi.spyOn(window, 'fetch').mockResolvedValue(searchLaunchId(randomStr));
    fireEvent.click(button);

    const errorMessage2 = await waitFor(() => getByTestId('error-message'));
    expect(errorMessage2).toHaveTextContent('The Launch ID provided was not found.');
  });
});
