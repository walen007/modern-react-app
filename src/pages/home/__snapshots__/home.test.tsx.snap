// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Home /> > --- should render home page lazy 1`] = `
<body>
  <div>
    <section
      class="_searchBox_c9479e"
    >
      <input
        aria-placeholder="Enter Launch ID"
        class="_txtSearch_0b168a"
        data-testid="search-text"
        placeholder="Enter Launch ID"
        tabindex="5"
        type="search"
      />
      <button
        aria-busy="false"
        aria-disabled="false"
        class="_btnSearch_723963"
        data-testid="submit-button"
        tabindex="6"
      >
        Search
      </button>
    </section>
    <section
      class="_pastLaunches_79102b"
    >
      <h2
        class="_heading_9b5d47"
        tabindex="23"
      >
        Past launches
      </h2>
      <ul
        aria-description="List of SpaceX past launches"
        class="_launchList_e86419"
        data-testid="past-launch-list"
      >
        <li
          aria-description="SpaceX past launches list item"
          aria-live="polite"
          class="_launchItem_e86419"
          tabindex="25"
        >
          <div
            class="_launchName_e86419"
            tabindex="26"
          >
            Crew-5
          </div>
          <div
            class="_launchId_e86419"
            tabindex="27"
          >
            ID: 
            62dd70d5202306255024d139
          </div>
        </li>
        <li
          aria-description="SpaceX past launches list item"
          aria-live="polite"
          class="_launchItem_e86419"
          tabindex="28"
        >
          <div
            class="_launchName_e86419"
            tabindex="29"
          >
            Starlink 4-35 (v1.5)
          </div>
          <div
            class="_launchId_e86419"
            tabindex="30"
          >
            ID: 
            63161339ffc78f3b8567070c
          </div>
        </li>
        <li
          aria-description="SpaceX past launches list item"
          aria-live="polite"
          class="_launchItem_e86419"
          tabindex="31"
        >
          <div
            class="_launchName_e86419"
            tabindex="32"
          >
            Starlink 4-34 (v1.5)
          </div>
          <div
            class="_launchId_e86419"
            tabindex="33"
          >
            ID: 
            63161329ffc78f3b8567070b
          </div>
        </li>
      </ul>
    </section>
  </div>
</body>
`;
